src/rest/server/pamAuth.go

```diff
@@ -70,48 +79,123 @@ func IsAdminGroup(username string) bool {
+	// }
+	// return false
+
 }
 
 func PAMAuthenAndAuthor(r *http.Request, rc *RequestContext) error {
 
+	root, err := os.Open("/")
+	if err != nil {
+		return err
+	}
+	defer root.Close()
+	if err := syscall.Chroot("/host_root"); err != nil {
+		root.Close()
+		return err
+	}
+
+	
+	defer syscall.Chroot(".")
+	defer root.Chdir()
+
 	username, passwd, authOK := r.BasicAuth()
-	if authOK == false {
+	if authOK == false || username == "" {
 		glog.Errorf("[%s] User info not present", rc.ID)
-		return httpError(http.StatusUnauthorized, "")
-	}
+		api_key_header := r.Header.Get("X-API-Key")
+		
+		if api_key_header == "" {
+			glog.Errorf("[%s] Api Key not present", rc.ID)
+			return httpError(http.StatusUnauthorized, "")
+		}
 
-	glog.Infof("[%s] Received user=%s", rc.ID, username)
-
-	/*
-	 * mgmt-framework container does not have access to /etc/passwd, /etc/group,
-	 * /etc/shadow and /etc/tacplus_conf files of host. One option is to share
-	 * /etc of host with /etc of container. For now disable this and use ssh
-	 * for authentication.
-	 */
-	/* err := PAMAuthUser(username, passwd)
-	    if err != nil {
-			log.Printf("Authentication failed. user=%s, error:%s", username, err.Error())
-	        return err
-	    }*/
-
-	//Use ssh for authentication.
-	config := &ssh.ClientConfig{
-		User: username,
-		Auth: []ssh.AuthMethod{
-			ssh.Password(passwd),
-		},
-		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
-	}
-	_, err := ssh.Dial("tcp", "127.0.0.1:22", config)
-	if err != nil {
-		glog.Infof("[%s] Failed to authenticate; %v", rc.ID, err)
-		return httpError(http.StatusUnauthorized, "")
-	}
+		api_key_parts := strings.Split(api_key_header, " ")
+		if len(api_key_parts) != 2 {
+			glog.Errorf("[%s] Invalid Api Key Format", rc.ID)
+			return httpError(http.StatusUnauthorized, "")
+		}
+		username = api_key_parts[0]
+		api_key := api_key_parts[1]
+		glog.Infof("[%s] Received user=%s", rc.ID, username)
+		filename := fmt.Sprintf("/etc/rest_api_keys/%v.key", username)
+		user_key_file_cont, err := ioutil.ReadFile(filename)
+		user_key := strings.TrimSpace(string(user_key_file_cont))
+		if err != nil {
+			glog.Errorf("[%s] Invalid Api Key User", rc.ID)
+			return httpError(http.StatusUnauthorized, "")
+		}
 
-	glog.Infof("[%s] Authentication passed. user=%s ", rc.ID, username)
+		if api_key != string(user_key) {
+
+			glog.Infof("[%s] Failed to authenticate, invalid API Key", rc.ID)
+			return httpError(http.StatusUnauthorized, "")
 
+		}
+	} else {
+
+		glog.Infof("[%s] Received user=%s", rc.ID, username)
+		/*
+		 * mgmt-framework container does not have access to /etc/passwd, /etc/group,
+		 * /etc/shadow and /etc/tacplus_conf files of host. One option is to share
+		 * /etc of host with /etc of container. For now disable this and use ssh
+		 * for authentication.
+		 */
+		err := PAMAuthUser(username, passwd)
+			if err != nil {
+				glog.Infof("[%s] Authentication failed. user=%s, error:%s", rc.ID, username, err.Error())
+				return err
+			}
```

Comment:
Can you check with tacacs+ authentication?

Reply:
I can try with tacas, I need to find a test setup for this. As far as @sachinholla #2 about openapi spec, since this is generated at build time using the pyang plugin, we would need to disable/enable auth at build time with some flag. Previously, the security section was not being defined at all in the openapi spec files. I statically added basicAuth and apiKey. If we need the ability to enable/disable these (plus client cert) we need some build flag I think.

Reply:
Authentication mode cannot be set at build time. We cannot have different builds just for authentication mode.

Reply:
For swagger API spec, you are supposed to set the securityDefinitions which defines the modes of authentication used, and then security section defines which parts of the API use which authentication types. I guess if we don't define these, but still enforce them in the rest server it can still work, there will just be no indication in the swagger UI that you need to authenticate a certain way. Lets have a meeting to discuss how we should proceed.